app/services/school_partnerships/assign_training_periods.rb

```diff
@@ -0,0 +1,34 @@
+module SchoolPartnerships
+  class AssignTrainingPeriods
+    def initialize(school_partnership:, school:, lead_provider:, contract_period:)
+      @school_partnership = school_partnership
+      @school = school
+      @lead_provider = lead_provider
+      @contract_period = contract_period
+    end
+
+    def call
+      TrainingPeriods::Search.new
+        .linkable_to_school_partnership(
+          school: @school,
+          lead_provider: @lead_provider,
+          contract_period: @contract_period
+        )
+        .find_each do |tp|
+          tp.update!(school_partnership: @school_partnership)
+
+          Events::Record.record_training_period_assigned_to_school_partnership_event!(
+            author: Events::LeadProviderAPIAuthor.new(lead_provider: @lead_provider),
```

Comment:
are we sure this service will always be used by providers via the API, maybe the create might be invoked by the admin interface?